### ISSUE TITLE
Matrix translation / rotation?

### ISSUE THREAD

**Issue by aneilbaboo** [josdejong/mathjs#1160]
What is the right way to translate or rotate a sparse or dense matrix?

**Comment by josdejong** [comment]
Right now there is only `transpose` and `ctranspose`. 

A translate/rotate function sounds interesting, we don't have that in mathjs. I guess we can get inspiration from similar functions in python/matlab/etc.

Do you have some concrete use cases for rotating a matrix?

**Comment by ericman314** [comment]
Typically rotation matrices are used to manipulate points or objects in space or on the plane, for example when doing 3D graphics. The matrix is constructed using trig functions, then multiplied to each vector in the object to be rotated. This allows for rotation of many points at a time. It would be very handy to have a set of functions to create rotation matrices. 

Or possibly @aneilbaboo means to rotate the elements themselves within a square matrix or something. I'm not familiar with the applications of doing that.

**Comment by aneilbaboo** [comment]
Thanks, @ericman314 - yes, that is what I meant.

**Comment by josdejong** [comment]
Ah, of course. Sounds cool!

**Comment by YuvalGat** [comment]
I think I could work on this. 
How about the following syntax?
```js
const vec = [1, 2];
const rotatedVec = math.rotate(vec, math.pi / 4); // [-0.707..., 2.12...]
```

**Comment by ericman314** [comment]
Could there also be a function that returns a rotation matrix? Then you could do the trig just once and rotate lots of vectors quickly:

```js
const vec = [1, 2];
const rotationMat = math.rotationMatrix(math.pi / 4); // [[0.707, -0.707], [0.707, 0.707]]

const rotatedVec = math.multiply(rotationMat, vec);   // [-0.707, 2.12]
```

I think 3D rotations are also a must-have, but I don't know about n-dimensional. I can't wrap my head around that!

**Comment by josdejong** [comment]
> I think I could work on this.

Thanks for your offer @YuvalGat, that's great!

I like the idea of @ericman314 , so that means we will create two new functions, where `math.rotate` is simply eye candy for `math.multiply(vec, math.rotationMatrix(angle))`. 

I also guess we need support for both 2D and 3D matrices? The function `math.rotate` can automatically detect this, but the functions `math.rotationMatrix` not. Maybe create two versions of the function: `math.rotationMatrix2D` and `math.rotationMatrix3D`?

Are there any lessons we can learn for the API from other languages/libraries?

**Comment by YuvalGat** [comment]
This sounds like a good idea! I'll get working ASAP.
I'll submit a PR when I have the basic 2D behaviour working properly. Still reading the code to understand how everything works, I believe I'll get it done within the next couple of days.

**Comment by rnd-debug** [comment]
@josdejong How dead is this issue? Looks like @YuvalGat gave up (?) and the basic `rotationMatrix(theta)` seems to be easy enough for me :)

**Comment by josdejong** [comment]
I think Yuval didn't manage to find time for it. Would be great if you can pick it up @rnd-debug, thanks!

**Comment by rnd-debug** [comment]
@josdejong ok, work started. Short question: in case of `rotate(theta, v)`, we can calculate the rotation matrix:
- coefficient-wise by manually calculating every coefficient. Like `R = [[ cos + vx*vx*(1-cos), etc.] etc.] `  => more wordy.
- OR by using the "Rodrigues' rotation formula" : `R = I + sin.K + (1-cos).K^2` => more mathjs dependencies need to be included (matrix multiplication, addition, identity). Probably slower and more prone to rounding issues.
What would be preferable? I am tending towards coefficient-wise as more standalone solution.
Using [this](https://en.wikipedia.org/wiki/Rotation_matrix#Rotation_matrix_from_axis_and_angle) as reference.

**Comment by josdejong** [comment]
Are there any downsides of using coefficient-wise?

**Comment by rnd-debug** [comment]
Overhead with typing the coefficients. :D so let's go for coefficient-wise.

**Comment by josdejong** [comment]
👍

**Comment by josdejong** [comment]
Function `rotationMatrix` is now available in `v7.4.0`, and @rnd-debug plans on creating `rotate` on top of that next. Let's keep this issue open until then.

**Comment by josdejong** [comment]
The new function `rotate` is now available in `v7.6.0`. Thanks again @rnd-debug . Closing this issue now.
